docs/source/workflow/generate-types.mdx

```diff
@@ -0,0 +1,205 @@
+---
+title: Generating types from a GraphQL schema
+description: How to ensure your resolvers are type safe
+---
+
+> 👋 If you haven't set up a TypeScript project using Apollo Server yet, follow our [Getting Started](../../getting-started) guide before continuing. 
+
+GraphQL uses a type system to clearly define the available data for each type and field in a [GraphQL schema](../schema/schema). Type generation libraries can take advantage of the strongly-typed nature of a GraphQL schema to automatically generate TypeScript types based on that schema. 
+
+You can use these generated TS types in your resolvers to type-check that your resolvers' return values match the field types dictated by your schema. Type checking your resolvers enables you to catch errors quickly and gives you the peace of mind that type safety ensures.
+
+> Looking to generate types for your [Apollo Federation](/federation/) subgraphs? Our [Subgraph template](https://github.com/apollographql/subgraph-template-typescript-apollo-server) lays down the groundwork so you can quickly set up a subgraph with generated types.
+
+## Setting up your project
+
+We'll use the [GraphQL Code Generator](https://www.the-guild.dev/graphql/codegen) library to generate types based on our GraphQL schema. There are [multiple ways](https://www.the-guild.dev/graphql/codegen/plugins/typescript/typescript-resolvers#usage-example) to provide a schema to GraphQL Code Generator. Below, we'll show one of the common methods, beginning by wrapping our schema in a `gql` template literal tag from the [`graphql-tag`](https://www.npmjs.com/package/graphql-tag) package.
+
+If you haven't already, begin by installing the `graphql-tag` library:
+
+```bash
+npm install graphql-tag
+```
+
+The `gql` template literal tag from `graphql-tag` can parse a Schema Definition Language (SDL) string into an abstract syntax tree (AST). Apollo Server accepts a string or an AST as a valid option for defining a schema.  
+
+If we pass GraphQL Code Generator a TypeScript file path, it will automatically generate types for the `gql` tag within that file. So, we can wrap our schema string in a `gql` tag to let GraphQL Code Generator know it should generate types for that tag. 
+
+Import the `gql` tag into the file where you define your schema, then wrap your schema string like so:
+
+```ts title="src/schema.ts"
```

Comment:
We can also provide plain `.graphql` files (when not dynamic definition is required)

Reply:
That was our original approach but .graphql files in TypeScript are a bit less simple since they aren't compiled or copied into dist. This also allows us to import the schema rather than read it from a file. The file copying can obviously be fixed with scripts or a build pipeline but we wanted to avoid that complexity to keep the example simple.

Reply:
Heh, I had written this as a pending comment:

> Do we really want to recommend putting the schema in a gql tag in a TS file here? Why not just a schema.graphql file (with `fs.readFileSync()` in the setup code)? I feel like a schema is likely to be long enough that it's worth being a standalone file...

I'm not sure I follow though — for an app (not a published npm package) why does it matter if a file is copied into dist?

Reply:
Good point that the publishing constraint is not a thing for an app. In that case, we'd want a separate folder adjacent to src and dist which a file from either could correctly reference? Having it in src wouldn't make much sense.

I'm missing the standalone file argument though, that's still the case with the gql tagged schema in schema.ts.

Reply:
Could just be a top level file and use `..`?

Idk, I think there are plenty of tools you might want to pass a schema to (like `rover`!) and having to use a TS parser to extract your schema when most likely it will be 99% of the file doesn't seem like the best pattern to push in our docs.

Reply:
(like, i know `rover subgraph introspect` exists, but just passing a schema file directly to `rover subgraph publish` seems easier in CI.)

Reply:
I'm convinced 👍

Sorry I led you astray @rkoron007!

Reply:
No problem, updating this back to using a top-level `schema.graphql` file!